crates/kernel/src/device/bcm2835_aux.rs

```diff
@@ -42,67 +119,126 @@ impl MiniUart {
         unsafe {
             // Enable mini UART, disable both SPI modules
             this.reg(Self::AUX_ENABLES).write(0b001);
-            // Clear bit 7, DLAB (baudrate register access instead of data)
-            // Clear bit 6, break
-            // Set bit 0, 8-bit mode
-            #[allow(clippy::eq_op)]
+
+            this.reg(Self::AUX_MU_IER_REG).write(IerReg::empty().bits());
+            this.reg(Self::AUX_MU_CNTL_REG)
+                .write(CntrlReg::empty().bits());
             this.reg(Self::AUX_MU_LCR_REG)
-                .write((0 << 7) | (0 << 6) | (1 << 0));
+                .write((LcrReg::DATA_8BIT).bits());
+            this.reg(Self::AUX_MU_MCR_REG).write(McrReg::empty().bits());
+            this.reg(Self::AUX_MU_IER_REG).write(IerReg::empty().bits()); // ????
+            this.reg(Self::AUX_MU_IIR_REG).write(0xC6); // ???
 
-            // TODO: initialize GPIO
+            let clock_rate;
+            {
+                let mut mailbox = super::MAILBOX.get().lock();
+                clock_rate = mailbox
+                    .get_property(super::mailbox::PropGetClockRate {
+                        id: super::mailbox::CLOCK_CORE,
+                    })
+                    .unwrap()
+                    .rate;
+            }
+
+            // baud_rate = sys_clock_freq / (8 * (baud_reg + 1))
+            // baud_reg = sys_clock_freq / (baud_rate * 8) - 1
+            let target_baud_rate = 115200;
+            let sys_clock_freq = clock_rate;
+            let baud_reg = sys_clock_freq / (target_baud_rate * 8) - 1;
+
+            this.reg(Self::AUX_MU_BAUD_REG).write(baud_reg);
+
+            {
+                let mut gpio = GPIO.get().lock();
+                gpio.set_function(14, GpioFunction::Alt5);
+                gpio.set_function(15, GpioFunction::Alt5);
```

Comment:
adding this is giving me output, although infrequent or slow or is just dropping a lot of characters/messages.
```suggestion

                gpio.set_pull(14, GpioPull::None);
                gpio.set_pull(15, GpioPull::None);
```

Reply:
ignore the - on line 154

Reply:
I'm assuming if you were able to get it working without this addition, then the bootloader set the pull to none for these pins?

Reply:
I would say try adding this on your end and see if it breaks anything